### ISSUE TITLE
-Wdangling support for containers with pointer-like elements

### ISSUE THREAD

**Issue by hokein** [llvm/llvm-project#100526]
It would be great if clang can diagnose the following cases:

```cpp
void test() {
  std::vector<std::string_view> t1;
  abc.push_back(std::string()); // now t1 holds a dangling pointer. #1


  std::optional<std::string_view> t2 = std::string(); // dangling  #2
  t2 = std::string();  // dangling
}
```

**Comment by chrchr-github** [comment]
`t1` is caught by `bugprone-dangling-handle`.
https://godbolt.org/z/qa6Werqxo

**Comment by vulptex** [comment]
still doesn't catch all cases:

https://godbolt.org/z/f9sWTqjvG

```c++
#include <string_view>
#include <string>
#include <vector>
#include <iostream>

void add(std::vector<std::string_view> &vec, int value) {
    auto str = std::to_string(value);
    vec.push_back(str);
}

int main(){

    std::vector<std::string_view> vec;
    add(vec, 1);
    addStruct(vec, 2);
    addStruct(vec, 3);

    for(auto& a: vec){
        std::cout << a << std::endl;
    }

    return 0;
}
```
